index/segment/fst/types.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fst
+
+import (
+	"io"
+	"regexp"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/postings"
+)
+
+const (
+	majorVersion = 1
+	minorVersion = 0
+)
+
+// Reader represents a FST segment.
+type Reader interface {
```

Comment:
Hm, is there a reason you don't embed `index.Reader` and instead call out the methods it implements explicitly?

Reply:
lol it's cause this type shares the same name :| Can't have interface X embed another interface X even if it's another package. Will rename this to Readable to avoid this issue.

Reply:
actually can't do that, cause both Reader and Segment have Close() which conflicts when embedding.

Reply:
Ugh.  Can you do something hacky like below to at least enforce it always meets the `index.Reader`?

Reply:
done

Reply:
What about calling it a `ReaderSegment` or `ReadableSegment`?